cfdm/data/abstract/mesharray.py

```diff
@@ -0,0 +1,366 @@
+
+            connectivity: array_like
+                A 2-d integer array of indices that corresponds to a
+                UGRID "edge_node_connectivity",
+                "face_node_connectivity", or
+                "face_face_connectivty" variable.
+
+            shape
+                The shape of the CF data model view of the
+                connectivity array.
+
+            {{int start_index: `int`}}
+
+            {{init cell_dimension: `int`}}
+
+            compression_type: `str`, optional
+                The type of compression.
+
+            {{init source: optional}}
+
+            {{init copy: `bool`, optional}}
+
+        """
+        super().__init__(
+            compressed_array=connectivity,
+            shape=shape,
+            compressed_dimensions=compressed_dimensions,
+            compression_type=compression_type,
+            source=source,
+            copy=copy,
+        )
+
+        if source is not None:
+            try:
+                start_index = source.get_start_index(None)
+            except AttributeError:
+                start_index = None
+
+            try:
+                cell_dimension = source.get_cell_dimension(None)
+            except AttributeError:
+                cell_dimension = None
+
+        if start_index is not None:
+            self._set_component("start_index", start_index, copy=False)
+
+        if cell_dimension is not None:
+            self._set_component("cell_dimension", cell_dimension, copy=False)
+
+    def __getitem__(self, indices):
+        """Return a subspace of the uncompressed data.
+
+        x.__getitem__(indices) <==> x[indices]
+
+        Returns a subspace of the uncompressed array as an independent
+        numpy array.
+
+        .. versionadded:: (cfdm) UGRIDVER
+
+        """
+        from math import isnan
+
+        # ------------------------------------------------------------
+        # Method: Uncompress the entire array and then subspace it
+        # ------------------------------------------------------------
+        # Initialise the un-sliced uncompressed array
+        shape = self.shape
+        known_shape = not any(map(isnan, shape))
+        if known_shape:
+            u = np.ma.empty(self.shape, dtype=self.dtype)
+
+        Subarray = self.get_Subarray()
+        subarray_kwargs = {
+            **self.conformed_data(),
+            **self.subarray_parameters(),
+        }
+
+        for u_indices, u_shape, c_indices, _ in zip(*self.subarrays()):
+            subarray = Subarray(
+                indices=c_indices,
+                shape=u_shape,
+                **subarray_kwargs,
+            )
+            if known_shape:
+                u[u_indices] = subarray[...]
+            else:
+                # When the shape is unknown, there must be (and so we
+                # are assuming that there is) only one pass through
+                # this loop.
+                u = subarray[...]
+
+        if not known_shape:
```

Comment:
```suggestion
        if known_shape:
```

Either this or the variable `known_shape` is confusingly named - it's hard to tell which contextually and when I run the full test suite with this change made, it still passes, so perhaps you could add a brief comment regardless to describe what 'known' in `known_shape` specifically means?

Reply:
`known_shape` is True if the shape of the uncompressed array is known _before_ uncompressing it. It will be False if the array represents node-node connectivity and the underlying data comes from a UGRID file. In this case, UGRID doesn't explicitly store this type of connectivity, so we don't known how many nodes are connected to any given node, i.e. the shape of the connectivity is unknown until we do the calculation at compute time.

I have add some comment that will hopefully clarify this: e34a190ef54e82f54469f7fb63731dad0e2db122

Reply:
Hi David, thanks for clarifying this. That's useful to know, for sure, so your new commit is good.

Though specifically (I probably should have been clearer) my immediate comment is questioning the fact that, given the name of the variable is `known_shape`, this little block seems to contradict itself (e.g. when read as pseudo-code on the `if` statement), which confuses me. Namely, `if not known_shape` indicates it is catching when the shape is _not_ known, but `# Store the shape for future refernce, now that is it known.` implies it _is_ known at this point. So it is confusion from the `knwon_shape` variable name, really.

But having read your explanation, the previous block where `u` is assigned and therefore the shape becoming known makes it clear enough, so it should be OK. as-is. I guess I didn't look at the slightly wider context which makes it make sense :slightly_smiling_face: Maybe to be fully clear you could add something brief like `# now known!` after `if known_shape:` so nobody will get a bit confused like I did, here?

Reply:
Hi Sadie - Ah ha. How about: 594f36e4e41248e2a595287312b553b64dd25593

Reply:
Thanks. Should be crystal clear now. Marking as resolved.